CMake/CheckHash.cmake

```
@@ -0,0 +1,35 @@
+# Validate md5 hash given md5file url and the file directory.
+
+# This module does the following on hash failure:
+# Set HASH_CHECK_FAIL to 1.
+# Remove the downloaded files.
+
+macro (check_hash MD5_URL DIR HASH_CHECK_FAIL)
+  set(HASH_CHECK_FAIL 0)
+  file(DOWNLOAD ${MD5_URL}
+      "${DIR}/hash.md5"
+      STATUS MD5_DOWNLOAD_STATUS_LIST)
+  list(GET MD5_DOWNLOAD_STATUS_LIST 0 MD5_DOWNLOAD_STATUS)
+  if (MD5_DOWNLOAD_STATUS EQUAL 0)
+    file(STRINGS "${DIR}/hash.md5" HASH_DATA NEWLINE_CONSUME)
+    string(REGEX REPLACE "\n" ";" HASH_LIST "${HASH_DATA}")
+    foreach(item ${HASH_LIST})
+      string(SUBSTRING ${item} 0 32 EXPECTED_HASH)
+      string(SUBSTRING ${item} 34 -1 FILE_NAME)
+      file(MD5 "${DIR}/${FILE_NAME}" LOCAL_HASH)
+      if (NOT LOCAL_HASH STREQUAL EXPECTED_HASH)
+        set(HASH_CHECK_FAIL 1)
+        file(REMOVE_RECURSE ${DIR})
+        message(WARNING
+            "Could not verify md5 hash!")
```

I think we can be a little clearer in this error message:

```suggestion
            "md5sum verification error for $(item)!  Got $(LOCAL_HASH), expected $(EXPECTED_HASH).")
```

---

Changed. A minor thing, it's `${` not `$(`.

---

Ah, yeah, thanks for the catch.  I have been writing Julia these days where the string substitution operator is `$()` not `${}`... so every time I switch to CMake or bash I get myself confused... :)